File: create_chunks.ts
import Python from 'tree-sitter-python';
import TreeSitter from 'tree-sitter';
import * as fs from 'fs';
import argparse from 'argparse';
import * as glob from 'glob';

const pyParser = new TreeSitter();
pyParser.setLanguage(Python);

interface Chunk {
    file: string;
    type: string;
    name: string | null;
    comments: string | null;
    text: string;
}

function parseFile(filePath: string) {
    const code = fs.readFileSync(filePath, 'utf8');
    return pyParser.parse(code);
}

function getName(node): string | null {
    const nameNode = node.childForFieldName('name');
    return nameNode ? nameNode.text : null;
}

function getChunks(filePath: string, maxChunkSize: number = 1000): Chunk[] {
    const tree = parseFile(filePath);
    const chunks: Chunk[] = [];
    let comment: string | null = null;
    let groups: string[] = [];
    let groupLength = 0;

    for (const node of tree.rootNode.children) {
        if (node.type === 'comment') {
            if (node.text === null) continue;
            if (comment === null) {
                comment = node.text;
            } else {
                comment += '\n' + node.text;
            }
            continue;
        }

        if (node.type === 'function_definition' || node.type === 'class_definition' || node.type === 'decorated_definition') {
            chunks.push({
                file: filePath,
                type: 'group',
                name: null,
                comments: null,
                text: groups.join('\n'),
            });
            groups.length = 0;
            groupLength = 0;
            chunks.push({
                file: filePath,
                type: node.type,
                name: getName(node),
                comments: comment,
                text: node.text,
            });
        } else {
            if (groupLength + node.text.length > maxChunkSize) {
                chunks.push({
                    file: filePath,
                    type: 'group',
                    name: null,
                    comments: null,
                    text: groups.join('\n'),
                });
                groups.length = 0;
                groupLength = 0;
            } else {
                let group = node.text;
                if (comment !== null) {
                    group = `${comment}\n${node.text}`;
                }
                groups.push(group);
                groupLength += node.text.length;
            }
        }
        comment = null;
    }
    return chunks;
}

if (true) {
    const parser = new argparse.ArgumentParser();
    parser.add_argument('--input_pattern', { type: String, required: true });
    parser.add_argument('--max_chunk_size', { type: Number, default: 1000 });
    parser.add_argument('--output', { type: String, default: 'chunks.json' });
    const args = parser.parse_args();

    const chunks: Chunk[] = [];
    const files = glob.sync(args.input_pattern, { nodir: true });
    for (const filePath of files) {
        console.log("File: ", filePath);
        chunks.push(...getChunks(filePath, args.maxChunkSize));
    }
    console.log(`Total Files Scanned: ${files.length}\tTotal Chunks: ${chunks.length}`);
    fs.writeFileSync(args.output, JSON.stringify(chunks));
}

File: create_chunks.py
import tree_sitter_python as tspython
from tree_sitter import Language, Parser
import pandas as pd
import argparse
import glob

PY_LANGUAGE = Language(tspython.language())
pyParser = Parser(PY_LANGUAGE)


def parse_file(file_path):
    with open(file_path, "r", encoding="utf-8") as file:
        code = file.read()
    tree = pyParser.parse(bytes(code, "utf8"))
    return tree


def get_name(node):
    if node.child_by_field_name("name") is not None:
        node.child_by_field_name("name").text
    else:
        return None


def get_chunks(file_path, max_chunk_size=1000):
    tree = parse_file(file_path)
    chunks = []
    comment = None
    groups = []
    group_length = 0
    for node in tree.root_node.children:
        if node.type == "comment":
            if node.text is None:
                continue
            if comment is None:
                comment = node.text.decode("utf8")
            else:
                comment = comment + "\n" + node.text.decode("utf8")
            continue

        if node.type in [
            "function_definition",
            "class_definition",
            "decorated_definition",
        ]:
            chunks.append(
                {
                    "file": file_path,
                    "type": "group",
                    "name": None,
                    "comments": None,
                    "text": "\n".join(groups),
                }
            )
            groups = []
            group_length = 0
            chunks.append(
                {
                    "file": file_path,
                    "type": node.type,
                    "name": get_name(node),
                    "comments": comment,
                    "text": node.text.decode("utf8"),
                }
            )
        else:
            if group_length + len(node.text) > max_chunk_size:
                chunks.append(
                    {
                        "file": file_path,
                        "type": "group",
                        "name": None,
                        "comments": None,
                        "text": "\n".join(groups),
                    }
                )
                groups = []
                group_length = 0

            else:
                group = node.text.decode("utf8")
                if comment is not None:
                    group = f"{comment}\n{node.text.decode("utf8")}"
                groups.append(group)
                group_length = group_length + len(node.text)
        comment = None
    return chunks


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--input_pattern", type=str, required=True)
    parser.add_argument("--max_chunk_size", type=int, default=1000)
    parser.add_argument("--output", type=str, default="chunks.parquet")
    args = parser.parse_args()

    chunks = []
    for path in glob.glob(args.input_pattern, recursive=True):
        chunks += get_chunks(path, args.max_chunk_size)

    pd.DataFrame(chunks).to_parquet(args.output)
